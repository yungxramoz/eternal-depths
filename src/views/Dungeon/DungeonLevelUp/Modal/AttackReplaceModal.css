.attack-list {
  display: flex;
  flex-direction: column;
  padding: 1em 0;
}
