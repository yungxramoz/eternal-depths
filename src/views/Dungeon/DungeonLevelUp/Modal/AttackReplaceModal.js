import React, { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import RpgButton from '../../../../components/atoms/RpgButton/RpgButton'
import Modal from '../../../../components/organisms/Modal/Modal'
import { characterReplaceAttack } from '../../../../store/game/gameSlice'
import ItemRadioButton from '../../../../components/organisms/ItemRadioButton/ItemRadioButton'

const AttackReplaceModal = ({ newAttack, onClose, onRewardSelected }) => {
  const dispatch = useDispatch()
  const attacks = useSelector((state) => state.game.character.current.attacks)
  const [selectedAttackId, setSelectedAttackId] = useState(2)
  const filteredAttacks = attacks.filter((attack) => attack.id !== 1)

  const handleReplace = () => {
    dispatch(characterReplaceAttack({ newAttack, attackId: selectedAttackId }))
    onRewardSelected()
  }

  return (
    <Modal
      title="Replace Attack"
      onDismiss={onClose}
      content={
        <div className="attack-replace-modal">
          <p>Select an attack to replace:</p>
          <form className="attack-list">
            {filteredAttacks.map((attack) => (
              <React.Fragment key={attack.id}>
                <ItemRadioButton
                  id={attack.id}
                  checked={selectedAttackId === attack.id}
                  title={attack.name}
                  subtitle={`Cooldown: ${attack.cooldown}`}
                  description={attack.description}
                  onChange={() => setSelectedAttackId(attack.id)}
                />
              </React.Fragment>
            ))}
          </form>
        </div>
      }
      footerContent={
        <RpgButton
          text="Replace"
          disabled={selectedAttackId == null}
          onClick={handleReplace}
        />
      }
    />
  )
}

export default AttackReplaceModal
