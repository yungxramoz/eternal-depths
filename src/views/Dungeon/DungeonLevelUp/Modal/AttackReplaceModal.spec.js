import { configureStore } from '@reduxjs/toolkit'
import { fireEvent, render, screen } from '@testing-library/react'
import React from 'react'
import { Provider } from 'react-redux'
import gameReducer from '../../../../store/game/gameSlice'
import AttackReplaceModal from './AttackReplaceModal'

describe('AttackReplaceModal', () => {
  let store
  let onRewardSelected

  beforeEach(() => {
    store = configureStore({ reducer: { game: gameReducer } })
    onRewardSelected = jest.fn()
  })

  it('renders correctly', () => {
    render(
      <Provider store={store}>
        <AttackReplaceModal
          newAttack={{}}
          onRewardSelected={onRewardSelected}
        />
      </Provider>,
    )
    expect(document.body).toMatchSnapshot('AttackReplaceModal')
  })

  it('dispatches characterReplaceAttack action and calls onRewardSelected when Replace button is clicked', () => {
    const dispatchSpy = jest.spyOn(store, 'dispatch')
    render(
      <Provider store={store}>
        <AttackReplaceModal
          newAttack={{}}
          onRewardSelected={onRewardSelected}
        />
      </Provider>,
    )
    fireEvent.click(screen.getByText('Replace'))
    expect(dispatchSpy).toHaveBeenCalledWith(
      expect.objectContaining({ type: 'game/characterReplaceAttack' }),
    )
    expect(onRewardSelected).toHaveBeenCalled()
  })
})
