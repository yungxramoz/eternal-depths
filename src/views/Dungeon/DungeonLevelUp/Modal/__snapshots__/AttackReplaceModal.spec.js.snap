// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttackReplaceModal renders correctly: AttackReplaceModal 1`] = `
<body>
  <div>
    <div
      class="modal-container"
    >
      <div
        class="rpgui-container container framed framed-golden modal"
      >
        <div
          class="modal-header"
        >
          <button
            class="icon-button large dismiss-button"
          >
            <svg
              fill="none"
              viewBox="0 0 24 24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M5 5h2v2H5V5zm4 4H7V7h2v2zm2 2H9V9h2v2zm2 0h-2v2H9v2H7v2H5v2h2v-2h2v-2h2v-2h2v2h2v2h2v2h2v-2h-2v-2h-2v-2h-2v-2zm2-2v2h-2V9h2zm2-2v2h-2V7h2zm0 0V5h2v2h-2z"
                fill="currentColor"
              />
            </svg>
          </button>
          <h2>
            Replace Attack
          </h2>
          <hr
            class="separator"
          />
        </div>
        <div
          class="modal-body"
        >
          <div
            class="attack-replace-modal"
          >
            <p>
              Select an attack to replace:
            </p>
            <form
              class="attack-list"
            >
              <input
                checked=""
                class="rpgui-radio golden"
                id="2"
                name="radio-golden"
                type="radio"
                value="2"
              />
              <label
                class="item-radio-label"
                for="2"
              >
                <div
                  class="item-container"
                >
                  <div
                    class="item-content"
                  >
                    <p
                      class="item-title"
                    >
                      Test Attack 2
                    </p>
                    <p
                      class="item-subtitle"
                    >
                      <i>
                        Cooldown: 1
                      </i>
                    </p>
                    <p
                      class="item-description"
                    >
                      Test Description 2
                    </p>
                  </div>
                </div>
              </label>
              <input
                class="rpgui-radio golden"
                id="3"
                name="radio-golden"
                type="radio"
                value="3"
              />
              <label
                class="item-radio-label"
                for="3"
              >
                <div
                  class="item-container"
                >
                  <div
                    class="item-content"
                  >
                    <p
                      class="item-title"
                    >
                      Test Attack 3
                    </p>
                    <p
                      class="item-subtitle"
                    >
                      <i>
                        Cooldown: 1
                      </i>
                    </p>
                    <p
                      class="item-description"
                    >
                      Test Description 3
                    </p>
                  </div>
                </div>
              </label>
            </form>
          </div>
        </div>
        <div
          class="modal-footer"
        >
          <hr
            class="separator"
          />
          <button
            class="rpgui-button btn"
          >
            Replace
          </button>
        </div>
      </div>
      <div
        class="modal-background"
      />
    </div>
  </div>
</body>
`;
